tests/playbooks_manager/test_views.py

```diff
@@ -0,0 +1,35 @@
+# This file is a part of IntelOwl https://github.com/intelowlproject/IntelOwl
+# See the file 'LICENSE' for copying permission.
+
+from api_app.playbooks_manager.serializers import PlaybookConfigSerializer
+
+from .. import CustomAPITestCase
+
+
+class PlaybookAppViewsTestCase(CustomAPITestCase):
+    def test_get_playbook_config(self):
+        response = self.client.get("/api/get_playbook_configs")
+        self.assertEqual(response.status_code, 200)
+        self.assertNotEqual(response.json(), {})
+        self.assertDictEqual(
+            response.json(), PlaybookConfigSerializer.read_and_verify_config()
+        )
+
+    def test_analyze_multiple_observables(self):
+        playbooks_to_run = ["FREE_TO_USE_ANALYZERS"]
+        data = {
+            "observables": [["ip", "1.1.1.1"]],
+            "playbooks_requested": playbooks_to_run,
+        }
+        response = self.client.post(
+            "/api/playbook/analyze_multiple_observables", data, format="json"
+        )
+        response_json = response.json()
+        playbooks_running = response_json.get("playbooks_running")
+
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(playbooks_running, playbooks_to_run)
+
+    # There is still space for file playbooks left here.
+    # I will be implementing them after we have a standard
+    # file analyzer to ourselves.
```

Comment:
yeah the project seems incomplete without proper testing. While we still need to initialize frontend tests and that's already a big issue in IntelOwl, at least we need to properly test new things in the backend because the framework does provide everything to do that easily.

You could just take the most important functions you implemented in all the code and write tests that verify that the use cases that you wants to handle are correctly handled:
1 - hypothesis
2 - initialize required data/parameters
3 - call function which needs testing
4 - verifiy results

In tests you do not have to care about code replication. For instance, if a specific function would give you 5 different results based on the different parameters that you provide, you would need to do 5 different tests for that, with the code really similar between tests. Each test should live its own life and be autonomous from the other.

For instance, the `job_cleanup` you implemented handle a lot of different possible outcomes. You need to verify those with tests.

Reply:
also `stack_connectors` and `stack_analyzers` are good targets

Reply:
the `before_run` too, you changed it. You should verify all the possible outcomes. Etc

Reply:
hmm. I am thinking of writing another test case that uploads files first. I did verify stack_connectors and stack_analyzers manually. And before_run was functionally verified after I compiled together the free analyzer playbook. And hopefully, just running the free analyzer playbook should do because it has all possible outcomes since it runs all possible analyzers.

Reply:
> hmm. I am thinking of writing another test case that uploads files first

ok!

> I did verify stack_connectors and stack_analyzers manually.

Don't get me wrong. I am not saying that you have not verified this manually. I am confident about that.
But that's not the same thing of writing unit tests. Unit tests are repeatable and allow to verify that, in the future, with new changes, what you have done now won't work differently. It also allow to avoid to break things while doing new features. They are as much important as the code itself that you have written.

> And before_run was functionally verified after I compiled together the free analyzer playbook. And hopefully, just running the free analyzer playbook should do because it has all possible outcomes since it runs all possible analyzers.

```
        if self._config.type == "file":
            if not job.file:
                raise UnsupportedObservableException(
                    f"No file was given for {self._config.name} file analyzer"
                )

            elif job.file_mimetype not in self._config.supported_filetypes:
                self.report.update_status(status=self.report.Status.FAILED)
                raise UnsupportedObservableException(
                    f"Observable {self._job.file_name} is not supported"
                )

            else:
                logger.info(
                    f"{job.file_name} is supported by the analyzer {self._config.name}"
                )
                return

        elif job.observable_classification not in self._config.observable_supported:
            self.report.update_status(status=self.report.Status.FAILED)
            raise UnsupportedObservableException(
                f"Observable {job.observable_name} is not supported"
            )
```

these are possible outcomes. Tests should verify that if those conditions apply, also if the exception raise (`self.assertException`). What you are saying is that the `before_run` code is actually run by your test. That's good for coverage but we are not doing "unit tests" there. Unit tests should pick different function, independently from the rest of the code, and test them with all the possible parameters, and see if the outcomes are what are expected